#### Séries Temprais

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
from datetime import datetime
from pandas.plotting import register_matplotlib_converters

O conjunto de dados `AirPassengers` contém as informações de transportes de passageiros nos Estados Unidos entre os anos de 1949 e 1960.

In [7]:
dataset = pd.read_csv("data/AirPassengers.csv")
dataset.head()

Month  #Passengers
0  1949-01          112
1  1949-02          118
2  1949-03          132
3  1949-04          129
4  1949-05          121

Pelo método `head` que mostra o cabeçalho o arquivo, observa-se que o dataset contém o número de passageros a cada mês.

In [5]:
dataset.shape

(144, 2)

O atributo `shape` mostra que o dataset contém 144 entradas.

Primeiro, é necessário analisar os tipos de dados das colunas do dataset, utilizando o atributo `dtypes`.

In [8]:
print(dataset.dtypes)

Month          object
#Passengers     int64
dtype: object


O mês está como objeto o que não permite a utilização dele para as análises, para isso, essa coluna precisa ser convertida em um formato de string de data, ANO-MÊS, através do método `strptime` da classe `datetime`.

Para isso cria-se um parser, para receber cada objeto do tipo data e transformá-lo em uma string do formato necessário.

Então o conjunto de dados é reimportado, mas dessa vez com novos parâmetros:
* `parse_dates` - indica qual coluna é a coluna das datas;
* `index_col` - indica qual coluna será a nova coluna de índices;
* `date_parser` - indica qual variável armazena o date parser que será utilizado.

In [11]:
dateparse = lambda dates: datetime.strptime(dates, "%Y-%m")

dataset = pd.read_csv("data/AirPassengers.csv",
                      parse_dates=["Month"],
                      index_col="Month",
                      date_parser=dateparse
)

dataset.head()

#Passengers
Month                  
1949-01-01          112
1949-02-01          118
1949-03-01          132
1949-04-01          129
1949-05-01          121

In [12]:
dataset.index

DatetimeIndex(['1949-01-01', '1949-02-01', '1949-03-01', '1949-04-01',
               '1949-05-01', '1949-06-01', '1949-07-01', '1949-08-01',
               '1949-09-01', '1949-10-01',
               ...
               '1960-03-01', '1960-04-01', '1960-05-01', '1960-06-01',
               '1960-07-01', '1960-08-01', '1960-09-01', '1960-10-01',
               '1960-11-01', '1960-12-01'],
              dtype='datetime64[ns]', name='Month', length=144, freq=None)

Agora observa-se que a coluna de meses está sendo uitlizada como o índice dos dados. E o seu tipo é objeto do tipo `datetime`.

In [13]:
time_series = dataset["#Passengers"]

A variável time series é criada utilizando a coluna de número de passageiros do dataset.

É possível também, visualizar esses dados de todas as formas que já são conhecidas.

In [29]:
print(f"Escolhendo um ano e um mês:\n{time_series['1949-08']}")
print(f"\nEscolhendo uma data específica:\n{time_series[datetime(1949,9,1)]}")
print(f"\nUitilizando intervalos:\n{time_series['1949-12-01' : '1950-02-01']}")
print(f"\nDo início até uma data específica:\n{time_series[: '1949-03-01']}")
print(f"\nDe uma data específica até o final:\n{time_series['1960-10-01':]}")
print(f"\nUm ano inteiro:\n{time_series['1957':]}")
print(f"\nValor máximo e mínimo:\nMáximo: {max(time_series)}\nMínimo: {min(time_series)}")

Escolhendo um ano e um mês:
Month
1949-08-01    148
Name: #Passengers, dtype: int64

Escolhendo uma data específica:
136

Uitilizando intervalos:
Month
1949-12-01    118
1950-01-01    115
1950-02-01    126
Name: #Passengers, dtype: int64

Do início até uma data específica:
Month
1949-01-01    112
1949-02-01    118
1949-03-01    132
Name: #Passengers, dtype: int64

De uma data específica até o final:
Month
1960-10-01    461
1960-11-01    390
1960-12-01    432
Name: #Passengers, dtype: int64

Um ano inteiro:
Month
1957-01-01    315
1957-02-01    301
1957-03-01    356
1957-04-01    348
1957-05-01    355
1957-06-01    422
1957-07-01    465
1957-08-01    467
1957-09-01    404
1957-10-01    347
1957-11-01    305
1957-12-01    336
1958-01-01    340
1958-02-01    318
1958-03-01    362
1958-04-01    348
1958-05-01    363
1958-06-01    435
1958-07-01    491
1958-08-01    505
1958-09-01    404
1958-10-01    359
1958-11-01    310
1958-12-01    337
1959-01-01    360
1959-02-01    342
1959-03-01    